File: benchmarks/weird_captcha_gym/shared_runtime/server/incubator_graders/impossible_ecology.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "impossible_ecology"


def _round(value: float, digits: int = 6) -> float:
    return round(float(value) + 1e-12, digits)


def _point(value: Any, arena: dict[str, Any]) -> list[float]:
    if not isinstance(value, list) or len(value) != 2 or any(isinstance(item, bool) for item in value):
        raise ValueError("field point is malformed")
    point = [float(value[0]), float(value[1])]
    if not (math.isfinite(point[0]) and math.isfinite(point[1]) and 0 <= point[0] <= float(arena["width"]) and 0 <= point[1] <= float(arena["height"])):
        raise ValueError("field point leaves arena")
    return point


def _initial_organisms(contract: dict[str, Any]) -> dict[str, dict[str, Any]]:
    result: dict[str, dict[str, Any]] = {}
    fields = set(contract["fields"])
    for raw in contract["organisms"]:
        organism_id = str(raw["id"])
        responses = {str(key): float(value) for key, value in raw["responses"].items()}
        if not organism_id or organism_id in result or set(responses) != fields:
            raise ValueError("organism response contract is malformed")
        result[organism_id] = {
            "id": organism_id,
            "radius": float(raw["radius"]),
            "responses": responses,
            "x": float(raw["initial_position"][0]), "y": float(raw["initial_position"][1]),
            "vx": 0.0, "vy": 0.0, "captured": False,
        }
    if len(result) != 5:
        raise ValueError("coupled ecology must contain five organisms")
    return result


def _targets(contract: dict[str, Any]) -> dict[str, dict[str, Any]]:
    result = {str(item["organism_id"]): item for item in contract["targets"]}
    if len(result) != 5:
        raise ValueError("sanctuary contract is malformed")
    return result


def _resolve_obstacle(organism: dict[str, Any], next_point: list[float], obstacle: dict[str, Any]) -> list[float]:
    dx = next_point[0] - float(obstacle["center"][0]); dy = next_point[1] - float(obstacle["center"][1])
    size = math.hypot(dx, dy); minimum = float(obstacle["radius"]) + organism["radius"]
    if size >= minimum:
        return next_point
    ux, uy = (dx / size, dy / size) if size > 1e-9 else (1.0, 0.0)
    inward = organism["vx"] * ux + organism["vy"] * uy
    if inward < 0:
        organism["vx"] -= 1.55 * inward * ux
        organism["vy"] -= 1.55 * inward * uy
    return [float(obstacle["center"][0]) + ux * minimum, float(obstacle["center"][1]) + uy * minimum]


def _advance(organisms: dict[str, dict[str, Any]], targets: dict[str, dict[str, Any]], contract: dict[str, Any], active: bool, selected: str | None, lure: list[float]) -> None:
    controls, arena, obstacle = contract["controls"], contract["arena"], contract["obstacle"]
    for organism in organisms.values():
        if organism["captured"]:
            continue
        if active and selected:
            dx, dy = lure[0] - organism["x"], lure[1] - organism["y"]
            size = math.hypot(dx, dy)
            if size > 1e-9:
                acceleration = organism["responses"][selected]
                organism["vx"] += dx / size * acceleration
                organism["vy"] += dy / size * acceleration
        organism["vx"] *= float(controls["damping"])
        organism["vy"] *= float(controls["damping"])
        speed = math.hypot(organism["vx"], organism["vy"])
        if speed > float(controls["max_speed"]):
            organism["vx"] = organism["vx"] / speed * float(controls["max_speed"])
            organism["vy"] = organism["vy"] / speed * float(controls["max_speed"])
        next_point = _resolve_obstacle(organism, [organism["x"] + organism["vx"], organism["y"] + organism["vy"]], obstacle)
        low = float(arena["margin"]) + organism["radius"]
        high_x = float(arena["width"]) - float(arena["margin"]) - organism["radius"]
        high_y = float(arena["height"]) - float(arena["margin"]) - organism["radius"]
        if next_point[0] < low or next_point[0] > high_x:
            next_point[0] = max(low, min(high_x, next_point[0])); organism["vx"] *= -.35
        if next_point[1] < low or next_point[1] > high_y:
            next_point[1] = max(low, min(high_y, next_point[1])); organism["vy"] *= -.35
        organism["x"] = _round(next_point[0]); organism["y"] = _round(next_point[1])
        organism["vx"] = _round(organism["vx"]); organism["vy"] = _round(organism["vy"])
        target = targets[organism["id"]]
        capture_radius = float(target["radius"]) - organism["radius"] - float(controls["capture_margin"])
        if math.dist([organism["x"], organism["y"]], [float(value) for value in target["center"]]) <= capture_radius and math.hypot(organism["vx"], organism["vy"]) <= float(controls["capture_speed"]):
            organism["x"], organism["y"] = [float(value) for value in target["center"]]
            organism["vx"] = organism["vy"] = 0.0; organism["captured"] = True


def _snapshot(organisms: dict[str, dict[str, Any]]) -> list[dict[str, Any]]:
    return [{"id": item["id"], "position": [_round(item["x"], 3), _round(item["y"], 3)], "velocity": [_round(item["vx"], 3), _round(item["vy"], 3)], "captured": item["captured"]} for item in sorted(organisms.values(), key=lambda value: value["id"])]


def _snapshot_matches(claimed: Any, expected: list[dict[str, Any]], tolerance: float = .035) -> bool:
    if not isinstance(claimed, list) or len(claimed) != len(expected):
        return False
    for actual, wanted in zip(claimed, expected):
        if not isinstance(actual, dict) or actual.get("id") != wanted["id"] or actual.get("captured") is not wanted["captured"]:
            return False
        for field in ("position", "velocity"):
            values = actual.get(field)
            if not isinstance(values, list) or len(values) != 2 or any(isinstance(item, bool) or not math.isfinite(float(item)) for item in values):
                return False
            if math.dist([float(item) for item in values], wanted[field]) > tolerance:
                return False
    return True


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    challenge_id, task_id = str(ground_truth.get("challenge_id") or ""), str(ground_truth.get("task_id") or "")
    if str(payload.get("mechanic_id") or "") != MECHANIC_ID or str(ground_truth.get("mechanic_id") or "") != MECHANIC_ID:
        return {"graded": True, "passed": False, "feedback": "mechanic mismatch"}
    if not challenge_id or str(payload.get("challenge_id") or "") != challenge_id or str(public_state.get("challenge_id") or "") != challenge_id:
        return {"graded": True, "passed": False, "feedback": "stale challenge"}
    if not task_id or str(payload.get("task_id") or "") != task_id:
        return {"graded": True, "passed": False, "feedback": "task identity mismatch"}
    for field in ("arena", "fields", "organisms", "targets", "obstacle", "controls"):
        if public_state.get(field) != ground_truth.get(field):
            return {"graded": True, "passed": False, "feedback": f"public/private ecology {field} contract skew"}
    try:
        organisms = _initial_organisms(ground_truth)
        targets = _targets(ground_truth)
        arena = ground_truth["arena"]
        fields = [str(item) for item in ground_truth["fields"]]
    except (KeyError, TypeError, ValueError) as exc:
        return {"graded": True, "passed": False, "feedback": f"invalid ecology contract: {exc}"}

    events = payload.get("events")
    if not isinstance(events, list) or not (1 <= len(events) <= 5000):
        return {"graded": True, "passed": False, "feedback": "ecology transcript is missing or outside limits"}
    selected: str | None = None
    lure = [float(arena["width"]) / 2, float(arena["height"]) / 2]
    active = completed = terminal = False
    tick = field_selections = pointer_drags = calibration_runs = resets = 0

    for sequence, event in enumerate(events, start=1):
        if terminal:
            return {"graded": True, "passed": False, "feedback": "interaction continued after all sanctuaries locked"}
        if not isinstance(event, dict) or event.get("sequence") != sequence:
            return {"graded": True, "passed": False, "feedback": f"event {sequence} sequence mismatch"}
        kind = str(event.get("kind") or "")
        if kind == "calibration":
            if active or event.get("tick") != tick or event.get("fields") != fields:
                return {"graded": True, "passed": False, "feedback": "calibration film event is malformed"}
            calibration_runs += 1
            continue
        if kind == "field_select":
            field = str(event.get("field") or "")
            if active or field not in fields or event.get("tick") != tick:
                return {"graded": True, "passed": False, "feedback": "field selection is malformed"}
            selected = field; field_selections += 1
            continue
        if kind in {"pointer_down", "pointer_move", "pointer_up"}:
            try:
                point = _point(event.get("point"), arena)
            except ValueError as exc:
                return {"graded": True, "passed": False, "feedback": f"event {sequence}: {exc}"}
            if event.get("tick") != tick or str(event.get("field") or "") != selected or selected not in fields:
                return {"graded": True, "passed": False, "feedback": f"event {sequence} uses stale field/tick state"}
            if kind == "pointer_down":
                if active:
                    return {"graded": True, "passed": False, "feedback": "pointer field was pressed twice"}
                active = True; pointer_drags += 1
            elif kind == "pointer_move":
                if not active:
                    return {"graded": True, "passed": False, "feedback": "field pointer moved while released"}
            else:
                if not active:
                    return {"graded": True, "passed": False, "feedback": "field pointer released while inactive"}
                active = False
            lure = point
            continue
        if kind == "reset":
            if active or completed:
                return {"graded": True, "passed": False, "feedback": "ecology reset during active/terminal field"}
            organisms = _initial_organisms(ground_truth); selected = None
            lure = [float(arena["width"]) / 2, float(arena["height"]) / 2]; tick = 0; resets += 1
            continue
        if kind == "physics_tick":
            if event.get("tick") != tick + 1 or tick >= int(ground_truth["controls"]["max_ticks"]):
                return {"graded": True, "passed": False, "feedback": "physics tick skipped, reversed, or exceeded"}
            if event.get("active") is not active or event.get("field") != (selected if active else None):
                return {"graded": True, "passed": False, "feedback": "physics tick field state disagrees with replay"}
            try:
                claimed_lure = _point(event.get("lure"), arena)
            except ValueError:
                return {"graded": True, "passed": False, "feedback": "physics tick lure is malformed"}
            if active and math.dist(claimed_lure, lure) > .01:
                return {"graded": True, "passed": False, "feedback": f"physics tick {tick + 1} lure moved without a pointer event: claimed={claimed_lure}, replay={lure}, event={sequence}"}
            if not active:
                # Hover motion is visually useful while inertia decays but has
                # no physical effect until the next recorded pointer-down.
                lure = claimed_lure
            tick += 1
            _advance(organisms, targets, ground_truth, active, selected, lure)
            if not _snapshot_matches(event.get("organisms"), _snapshot(organisms)):
                return {"graded": True, "passed": False, "feedback": f"physics tick {tick} fabricates coupled organism motion"}
            continue
        if kind == "complete":
            if completed or event.get("tick") != tick or not all(item["captured"] for item in organisms.values()):
                return {"graded": True, "passed": False, "feedback": "completion lacks five physically captured organisms"}
            completed = terminal = True
            continue
        return {"graded": True, "passed": False, "feedback": f"event {sequence} has unknown kind"}

    expected = _snapshot(organisms)
    if not _snapshot_matches(payload.get("final_organisms"), expected):
        return {"graded": True, "passed": False, "feedback": "submitted final organisms disagree with physics replay"}
    expected_scalars = {"tick": tick, "completed": completed, "field_selections": field_selections, "pointer_drags": pointer_drags, "calibration_runs": calibration_runs, "resets": resets}
    for field, value in expected_scalars.items():
        if payload.get(field) != value:
            return {"graded": True, "passed": False, "feedback": f"submitted {field} disagrees with ecology replay"}
    captured = sum(item["captured"] for item in organisms.values())
    passed = completed and captured == len(organisms) and pointer_drags > 0
    return {
        "graded": True,
        "passed": passed,
        "score": 100 if passed else 0,
        "feedback": f"coupled ecology replay: sanctuaries {captured}/{len(organisms)}; field selections {field_selections}; pointer interventions {pointer_drags}; ticks {tick}; resets {resets}",
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    del public_state
    return {"organisms": ground_truth.get("organisms") or [], "targets": ground_truth.get("targets") or [], "answers": []}

File: benchmarks/weird_captcha_gym/shared_runtime/app/mechanics/impossible_ecology.js
(() => {
  "use strict";

  let model = null;

  const esc = (value) => String(value == null ? "" : value).replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  const round = (value, digits = 6) => Math.round(value * (10 ** digits)) / (10 ** digits);
  const distance = (a, b) => Math.hypot(a[0] - b[0], a[1] - b[1]);

  function record(kind, details = {}) {
    const event = {sequence: model.events.length + 1, kind, ...details};
    model.events.push(event);
    return event;
  }

  function clearFresh() {
    const root = document.querySelector(".impossible-ecology-captcha");
    if (!root || root.dataset.freshFailure !== "true") return;
    root.dataset.freshFailure = "false";
    document.querySelector(".eco-fail-stamp")?.remove();
    model.helpers.setReadout("FIELD LAB ACTIVE · CALIBRATE OR INTERVENE", "idle");
  }

  function initialOrganisms() {
    return Object.fromEntries(model.state.organisms.map((item) => [item.id, {
      id: item.id,
      label: item.label,
      color: item.color,
      radius: Number(item.radius),
      responses: {...item.responses},
      x: Number(item.initial_position[0]),
      y: Number(item.initial_position[1]),
      vx: 0,
      vy: 0,
      captured: false,
      trail: [],
    }]));
  }

  function resetWorld(recordEvent = true) {
    if (!model || model.submitting || model.terminal) return;
    clearFresh();
    if (recordEvent) record("reset");
    model.organisms = initialOrganisms();
    model.tick = 0;
    model.active = false;
    model.pointerDown = false;
    model.completed = false;
    model.selectedField = null;
    model.lure = [model.state.arena.width / 2, model.state.arena.height / 2];
    model.resets += recordEvent ? 1 : 0;
    document.querySelector(".eco-complete")?.setAttribute("data-visible", "false");
    updateHUD();
    model.helpers.setReadout("ECOSYSTEM RESET · ALL ORGANISMS MOBILE", "idle");
  }

  function targetFor(id) {
    return model.targets.get(id);
  }

  function resolveObstacle(organism, next) {
    const obstacle = model.state.obstacle;
    const dx = next[0] - obstacle.center[0], dy = next[1] - obstacle.center[1];
    const size = Math.hypot(dx, dy);
    const minimum = Number(obstacle.radius) + organism.radius;
    if (size >= minimum) return next;
    const ux = size > 1e-9 ? dx / size : 1, uy = size > 1e-9 ? dy / size : 0;
    const inward = organism.vx * ux + organism.vy * uy;
    if (inward < 0) {
      organism.vx -= 1.55 * inward * ux;
      organism.vy -= 1.55 * inward * uy;
    }
    return [obstacle.center[0] + ux * minimum, obstacle.center[1] + uy * minimum];
  }

  function advancePhysics() {
    const controls = model.state.controls;
    const arena = model.state.arena;
    for (const organism of Object.values(model.organisms)) {
      if (organism.captured) continue;
      if (model.active && model.selectedField) {
        const dx = model.lure[0] - organism.x, dy = model.lure[1] - organism.y;
        const size = Math.hypot(dx, dy);
        if (size > 1e-9) {
          const acceleration = Number(organism.responses[model.selectedField]);
          organism.vx += dx / size * acceleration;
          organism.vy += dy / size * acceleration;
        }
      }
      organism.vx *= Number(controls.damping);
      organism.vy *= Number(controls.damping);
      const speed = Math.hypot(organism.vx, organism.vy);
      if (speed > Number(controls.max_speed)) {
        organism.vx = organism.vx / speed * Number(controls.max_speed);
        organism.vy = organism.vy / speed * Number(controls.max_speed);
      }
      let next = resolveObstacle(organism, [organism.x + organism.vx, organism.y + organism.vy]);
      const low = Number(arena.margin) + organism.radius;
      const highX = Number(arena.width) - Number(arena.margin) - organism.radius;
      const highY = Number(arena.height) - Number(arena.margin) - organism.radius;
      if (next[0] < low || next[0] > highX) { next[0] = Math.max(low, Math.min(highX, next[0])); organism.vx *= -.35; }
      if (next[1] < low || next[1] > highY) { next[1] = Math.max(low, Math.min(highY, next[1])); organism.vy *= -.35; }
      organism.x = round(next[0]); organism.y = round(next[1]); organism.vx = round(organism.vx); organism.vy = round(organism.vy);
      organism.trail.push([organism.x, organism.y]);
      if (organism.trail.length > 26) organism.trail.shift();
      const target = targetFor(organism.id);
      const captureRadius = Number(target.radius) - organism.radius - Number(controls.capture_margin);
      if (distance([organism.x, organism.y], target.center) <= captureRadius && Math.hypot(organism.vx, organism.vy) <= Number(controls.capture_speed)) {
        organism.x = Number(target.center[0]); organism.y = Number(target.center[1]);
        organism.vx = 0; organism.vy = 0; organism.captured = true;
      }
    }
  }

  function snapshot() {
    return Object.values(model.organisms).sort((a, b) => a.id.localeCompare(b.id)).map((item) => ({
      id: item.id,
      position: [round(item.x, 3), round(item.y, 3)],
      velocity: [round(item.vx, 3), round(item.vy, 3)],
      captured: item.captured,
    }));
  }

  function physicsTick() {
    if (!model || model.submitting || model.terminal || model.calibrating) return;
    const moving = Object.values(model.organisms).some((item) => !item.captured && Math.hypot(item.vx, item.vy) > .015);
    if (!model.active && !moving) return;
    if (model.tick >= Number(model.state.controls.max_ticks)) {
      model.active = false;
      model.pointerDown = false;
      model.helpers.setReadout("FIELD BUDGET EXHAUSTED · RESET OR SUBMIT", "error");
      return;
    }
    model.tick += 1;
    advancePhysics();
    record("physics_tick", {
      tick: model.tick,
      active: model.active,
      field: model.active ? model.selectedField : null,
      lure: model.lure.map((value) => round(value, 3)),
      organisms: snapshot(),
    });
    const captured = Object.values(model.organisms).filter((item) => item.captured).length;
    if (captured === Object.keys(model.organisms).length && !model.completed) {
      model.completed = true;
      model.active = false;
      model.pointerDown = false;
      record("complete", {tick: model.tick});
      document.querySelector(".eco-complete")?.setAttribute("data-visible", "true");
      model.helpers.setReadout("ALL FIVE SANCTUARIES STABLE · READY TO CERTIFY", "idle");
    }
    updateHUD();
  }

  function arenaPoint(event) {
    const box = model.canvas.getBoundingClientRect();
    return [
      Math.max(0, Math.min(model.state.arena.width, (event.clientX - box.left) / box.width * model.state.arena.width)),
      Math.max(0, Math.min(model.state.arena.height, (event.clientY - box.top) / box.height * model.state.arena.height)),
    ];
  }

  function selectField(field) {
    if (!model || model.active || model.submitting || model.terminal || model.calibrating || !model.state.fields.includes(field)) return;
    clearFresh();
    model.selectedField = field;
    model.fieldSelections += 1;
    record("field_select", {field, tick: model.tick});
    updateHUD();
    model.helpers.setReadout(`${field} FIELD ARMED · HOLD THE POINTER IN THE ARENA`, "idle");
  }

  function pointerDown(event) {
    if (!model?.selectedField || model.completed || model.submitting || model.terminal || model.calibrating) return;
    event.preventDefault();
    clearFresh();
    model.lure = arenaPoint(event);
    model.active = true;
    model.pointerDown = true;
    model.pointerDrags += 1;
    record("pointer_down", {tick: model.tick, field: model.selectedField, point: model.lure.map((value) => round(value, 3))});
    model.canvas.setPointerCapture?.(event.pointerId);
    model.helpers.setReadout(`${model.selectedField} FIELD LIVE · ALL MOBILE ORGANISMS RESPOND`, "pending");
  }

  function pointerMove(event) {
    if (!model) return;
    const point = arenaPoint(event);
    if (!model.pointerDown) { model.lure = point; return; }
    if (distance(point, model.lure) < Number(model.state.controls.pointer_sample_distance)) return;
    model.lure = point;
    record("pointer_move", {tick: model.tick, field: model.selectedField, point: model.lure.map((value) => round(value, 3))});
  }

  function pointerUp(event) {
    if (!model?.pointerDown) return;
    model.lure = arenaPoint(event);
    record("pointer_up", {tick: model.tick, field: model.selectedField, point: model.lure.map((value) => round(value, 3))});
    model.pointerDown = false;
    model.active = false;
    model.helpers.setReadout("FIELD RELEASED · INERTIA DECAYING", "idle");
  }

  function runCalibration() {
    if (!model || model.active || model.submitting || model.terminal || model.calibrating) return;
    clearFresh();
    model.calibrationRuns += 1;
    model.calibrating = {started: performance.now()};
    record("calibration", {tick: model.tick, fields: [...model.state.fields]});
    document.querySelector(".eco-calibrate")?.setAttribute("disabled", "disabled");
    model.helpers.setReadout("CALIBRATION FILM RUNNING · WATCH EVERY ORGANISM", "pending");
  }

  function calibrationState(now) {
    if (!model.calibrating) return null;
    const elapsed = now - model.calibrating.started;
    const duration = Number(model.state.controls.calibration_field_ms);
    const index = Math.floor(elapsed / duration);
    if (index >= model.state.fields.length) {
      model.calibrating = null;
      document.querySelector(".eco-calibrate")?.removeAttribute("disabled");
      model.helpers.setReadout("CALIBRATION COMPLETE · POSITIONS RESTORED · INTERVENE", "idle");
      return null;
    }
    return {field: model.state.fields[index], progress: (elapsed % duration) / duration};
  }

  function updateHUD() {
    const captured = Object.values(model.organisms).filter((item) => item.captured).length;
    const root = document.querySelector(".impossible-ecology-captcha");
    if (root) root.dataset.captured = String(captured);
    const count = document.getElementById("eco-capture-count");
    if (count) count.textContent = `${captured} / ${Object.keys(model.organisms).length}`;
    const tick = document.getElementById("eco-tick-count");
    if (tick) tick.textContent = String(model.tick).padStart(4, "0");
    document.querySelectorAll("[data-field]").forEach((button) => button.dataset.selected = button.dataset.field === model.selectedField ? "true" : "false");
    const ledger = document.getElementById("eco-organism-ledger");
    if (ledger) ledger.innerHTML = Object.values(model.organisms).map((item) => `<li data-captured="${item.captured}"><i style="--organism:${esc(item.color)}">${esc(item.label)}</i><span>${item.captured ? "SANCTUARY LOCKED" : "RESPONDING TO GLOBAL FIELD"}</span><b>${item.captured ? "STABLE" : "MOBILE"}</b></li>`).join("");
  }

  function draw(now) {
    if (!model) return;
    const ctx = model.canvas.getContext("2d");
    const {width, height} = model.state.arena;
    const palette = model.state.palette;
    ctx.clearRect(0, 0, width, height);
    ctx.fillStyle = palette.paper; ctx.fillRect(0, 0, width, height);
    ctx.strokeStyle = palette.grid; ctx.lineWidth = 1; ctx.globalAlpha = .42;
    for (let x = 20; x < width; x += 40) { ctx.beginPath(); ctx.moveTo(x, 0); ctx.lineTo(x, height); ctx.stroke(); }
    for (let y = 15; y < height; y += 40) { ctx.beginPath(); ctx.moveTo(0, y); ctx.lineTo(width, y); ctx.stroke(); }
    ctx.globalAlpha = 1;
    const obstacle = model.state.obstacle;
    ctx.beginPath(); ctx.arc(obstacle.center[0], obstacle.center[1], obstacle.radius, 0, Math.PI * 2); ctx.fillStyle = `${palette.danger}18`; ctx.fill(); ctx.setLineDash([8, 6]); ctx.strokeStyle = palette.danger; ctx.lineWidth = 2; ctx.stroke(); ctx.setLineDash([]); ctx.fillStyle = palette.danger; ctx.font = "800 9px ui-monospace"; ctx.textAlign = "center"; ctx.fillText("STERILE NURSERY / SOLID", obstacle.center[0], obstacle.center[1] + 3);
    for (const target of model.state.targets) {
      const organism = model.organisms[target.organism_id];
      ctx.beginPath(); ctx.arc(target.center[0], target.center[1], target.radius, 0, Math.PI * 2); ctx.fillStyle = `${target.color}12`; ctx.fill(); ctx.strokeStyle = target.color; ctx.lineWidth = organism.captured ? 5 : 2; ctx.setLineDash(organism.captured ? [] : [7, 6]); ctx.stroke(); ctx.setLineDash([]); ctx.fillStyle = target.color; ctx.font = "900 12px ui-monospace"; ctx.fillText(`SANCTUARY ${target.label}`, target.center[0], target.center[1] + 4);
    }
    const calibration = calibrationState(now);
    for (const organism of Object.values(model.organisms)) {
      if (organism.trail.length > 1) { ctx.beginPath(); organism.trail.forEach((point, index) => index ? ctx.lineTo(point[0], point[1]) : ctx.moveTo(point[0], point[1])); ctx.strokeStyle = `${organism.color}66`; ctx.lineWidth = 2; ctx.stroke(); }
      if (calibration) {
        const dx = model.state.obstacle.center[0] - organism.x, dy = model.state.obstacle.center[1] - organism.y, size = Math.hypot(dx, dy) || 1;
        const offset = Number(organism.responses[calibration.field]) * 27 * Math.sin(calibration.progress * Math.PI);
        ctx.beginPath(); ctx.arc(organism.x + dx / size * offset, organism.y + dy / size * offset, organism.radius + 5, 0, Math.PI * 2); ctx.strokeStyle = organism.color; ctx.globalAlpha = .45; ctx.setLineDash([4, 4]); ctx.stroke(); ctx.setLineDash([]); ctx.globalAlpha = 1;
      }
      ctx.beginPath(); ctx.arc(organism.x, organism.y, organism.radius, 0, Math.PI * 2); ctx.fillStyle = organism.color; ctx.shadowColor = organism.color; ctx.shadowBlur = organism.captured ? 18 : 8; ctx.fill(); ctx.shadowBlur = 0; ctx.fillStyle = "#07100a"; ctx.font = "900 13px ui-monospace"; ctx.fillText(organism.label, organism.x, organism.y + 5);
    }
    if (calibration) { ctx.fillStyle = `${palette.paper}dd`; ctx.fillRect(355, 16, 290, 45); ctx.strokeStyle = palette.ink; ctx.strokeRect(355, 16, 290, 45); ctx.fillStyle = palette.ink; ctx.font = "900 13px ui-monospace"; ctx.fillText(`CALIBRATION FILM · ${calibration.field}`, 500, 44); }
    else if (model.selectedField) { ctx.strokeStyle = "#ffffff"; ctx.lineWidth = 2; ctx.beginPath(); ctx.arc(model.lure[0], model.lure[1], 12, 0, Math.PI * 2); ctx.moveTo(model.lure[0] - 18, model.lure[1]); ctx.lineTo(model.lure[0] + 18, model.lure[1]); ctx.moveTo(model.lure[0], model.lure[1] - 18); ctx.lineTo(model.lure[0], model.lure[1] + 18); ctx.stroke(); ctx.fillStyle = palette.ink; ctx.font = "800 8px ui-monospace"; ctx.fillText(`${model.selectedField}${model.active ? " LIVE" : " ARMED"}`, model.lure[0], model.lure[1] - 22); }
    model.raf = requestAnimationFrame(draw);
  }

  async function submit() {
    if (!model || model.submitting || model.terminal) return;
    model.submitting = true;
    model.helpers.setReadout("REPLAYING COUPLED FIELD PHYSICS…", "pending");
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      events: model.events,
      final_organisms: snapshot(),
      tick: model.tick,
      completed: model.completed,
      field_selections: model.fieldSelections,
      pointer_drags: model.pointerDrags,
      calibration_runs: model.calibrationRuns,
      resets: model.resets,
    };
    try {
      const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify(payload)});
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.terminal = true;
        document.querySelector(".impossible-ecology-captcha")?.insertAdjacentHTML("beforeend", '<div class="eco-verdict"><span>COUPLED ECOSYSTEM STABLE</span><strong>PASS</strong><i>FIELD / MOTION / SANCTUARY REPLAY VERIFIED</i></div>');
        model.helpers.setReadout("PASS", "passed");
      } else if (outcome.passed === false && outcome.state) {
        await model.helpers.render(outcome.state);
        const root = document.querySelector(".impossible-ecology-captcha");
        root?.setAttribute("data-fresh-failure", "true");
        root?.insertAdjacentHTML("afterbegin", '<div class="eco-fail-stamp"><b>FAIL</b><span>UNSTABLE ECOLOGY · FRESH FIELD ISSUED</span></div>');
        const readout = document.querySelector(".readout"); if (readout) { readout.textContent = "FAIL · FRESH COUPLED FIELD ISSUED"; readout.dataset.status = "error"; }
      } else { model.submitting = false; model.helpers.setReadout("FAIL · ECOLOGY GRADE UNAVAILABLE", "error"); }
    } catch (_error) { model.submitting = false; model.helpers.setReadout("FAIL · ECOLOGY VERIFIER OFFLINE", "error"); }
  }

  async function render(state, helpers) {
    if (model?.raf) cancelAnimationFrame(model.raf);
    if (model?.timer) clearInterval(model.timer);
    document.body.dataset.mechanic = "impossible-ecology";
    document.body.dataset.cheatMode = helpers.isCheatMode() ? "true" : "false";
    helpers.app.innerHTML = `<section class="impossible-ecology-captcha" data-challenge-id="${esc(state.challenge_id)}" data-fresh-failure="false" data-captured="0" style="--eco-grid:${esc(state.palette.grid)};--eco-ink:${esc(state.palette.ink)};--eco-danger:${esc(state.palette.danger)}"><header class="eco-head"><div><span>IMPOSSIBLE ECOLOGY / COUPLED FIELD ${esc(state.challenge_id)}</span><h1>${esc(state.prompt)}</h1></div><aside><span>STABLE SANCTUARIES</span><b id="eco-capture-count">0 / 5</b><i>TICK <strong id="eco-tick-count">0000</strong></i></aside></header><main class="eco-main"><section class="eco-arena-shell"><canvas class="eco-arena" width="${state.arena.width}" height="${state.arena.height}"></canvas><div class="eco-arena-note">SELECT FIELD · HOLD / MOVE POINTER · RELEASE TO DAMP</div><div class="eco-complete" data-visible="false"><b>ALL SANCTUARIES STABLE</b><span>CERTIFY THE LIVING FIELD</span></div></section><aside class="eco-console"><p class="eco-console-label">GLOBAL FIELD CONSOLE</p><h2>One pointer. Five incompatible responses.</h2><div class="eco-fields">${state.fields.map((field, index) => `<button type="button" data-field="${esc(field)}"><i>${String(index + 1).padStart(2, "0")}</i><b>${esc(field)}</b><span>ARM FIELD</span></button>`).join("")}</div><button type="button" class="eco-calibrate">RUN THREE-FIELD CALIBRATION FILM</button><ol id="eco-organism-ledger" class="eco-ledger"></ol><div class="eco-rules">${state.rules.map((rule) => `<p>${esc(rule)}</p>`).join("")}</div></aside></main><footer class="eco-foot"><button type="button" class="eco-reset">RESET ECOSYSTEM</button><div><span>FIELD STATUS</span><div class="readout" data-status="idle">FIELD LAB ACTIVE · CALIBRATE OR INTERVENE</div></div><button type="button" class="eco-submit">${esc(state.submit_label)}</button></footer>${helpers.cheatPanelTemplate()}</section>`;
    model = {state, helpers, canvas: document.querySelector(".eco-arena"), targets: new Map(state.targets.map((target) => [target.organism_id, target])), organisms: {}, selectedField: null, lure: [state.arena.width / 2, state.arena.height / 2], active: false, pointerDown: false, tick: 0, completed: false, events: [], fieldSelections: 0, pointerDrags: 0, calibrationRuns: 0, resets: 0, calibrating: null, submitting: false, terminal: false, raf: null, timer: null};
    model.organisms = initialOrganisms();
    window.impossibleEcologyModel = model;
    document.querySelectorAll("[data-field]").forEach((button) => button.addEventListener("click", () => selectField(button.dataset.field)));
    document.querySelector(".eco-calibrate")?.addEventListener("click", runCalibration);
    document.querySelector(".eco-reset")?.addEventListener("click", () => resetWorld(true));
    document.querySelector(".eco-submit")?.addEventListener("click", submit);
    model.canvas.addEventListener("pointerdown", pointerDown);
    model.canvas.addEventListener("pointermove", pointerMove);
    model.canvas.addEventListener("pointerup", pointerUp);
    model.canvas.addEventListener("pointercancel", pointerUp);
    model.timer = setInterval(physicsTick, Number(state.controls.tick_ms));
    updateHUD();
    model.raf = requestAnimationFrame(draw);
    helpers.installCheatPanel();
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.impossible_ecology = {rootSelector: ".impossible-ecology-captcha", render};
})();
